common/flatpak-installation.c

```diff
@@ -2884,33 +2884,168 @@ flatpak_installation_run_triggers (FlatpakInstallation *self,
+  if (system_dir == NULL)
+    g_hash_table_add (used_refs, g_strdup (ref));
+
+  /* If @system_dir is non-NULL, that means @ref exists in @dir but we should
+   * look in @system_dir for related things */
+  if (system_dir != NULL)
+    related = flatpak_dir_find_local_related (system_dir, ref, dir, origin, TRUE, NULL, NULL);
+  else
+    related = flatpak_dir_find_local_related (dir, ref, NULL, origin, TRUE, NULL, NULL);
 
-  related = flatpak_dir_find_local_related (dir, ref, origin, TRUE, NULL, NULL);
   if (related == NULL)
     return;
 
   for (i = 0; i < related->len; i++)
     {
       FlatpakRelated *rel = g_ptr_array_index (related, i);
 
+      /* Check if this related ref is present in @dir, which implies the one
+       * in @system_dir is NOT the one being used. */
+      if (system_dir != NULL)
+        {
+          g_autoptr(FlatpakDeploy) user_related_deploy = flatpak_dir_load_deployed (dir, rel->ref, NULL, NULL, NULL);
+          if (user_related_deploy != NULL)
+            continue;
+        }
+
       if (!rel->auto_prune && !g_hash_table_contains (used_refs, rel->ref))
         {
           g_autofree char *related_origin = NULL;
 
           g_hash_table_add (used_refs, g_strdup (rel->ref));
 
-          related_origin = flatpak_dir_get_origin (dir, rel->ref, NULL, NULL);
+          related_origin = flatpak_dir_get_origin (system_dir ? system_dir : dir, rel->ref, NULL, NULL);
           if (related_origin != NULL)
-            find_used_refs (dir, used_refs, rel->ref, related_origin);
+            find_used_related_refs (system_dir ? system_dir : dir, NULL, used_refs, rel->ref, related_origin);
+        }
+    }
+}
+
+static void
+find_used_refs_for_apps (FlatpakDir  *dir,
+                         FlatpakDir  *system_dir, /* nullable */
+                         char       **app_refs,
+                         const char  *arch,
+                         GHashTable  *used_runtimes,
+                         GHashTable  *used_refs)
+{
+  /* Check for related refs and runtimes and sdks for each app in @app_refs.
+   * The apps exist in @dir but if @system_dir is set that's where we check for
+   * the related/runtime/sdk, and if @system_dir is set we check that said
+   * runtime does not exist in @dir, so the one in @system_dir is probably the
+   * one being used. */
+  int i;
+  for (i = 0; app_refs[i] != NULL; i++)
+    {
+      const char *ref = app_refs[i];
+      g_autoptr(FlatpakDeploy) deploy = NULL;
+      g_autofree char *origin = NULL;
+      g_autofree char *runtime = NULL;
+      g_autofree char *sdk = NULL;
+      g_autoptr(GKeyFile) metakey = NULL;
+      g_auto(GStrv) parts = g_strsplit (ref, "/", -1);
+
+      if (arch != NULL && strcmp (parts[2], arch) != 0)
+        continue;
+
+      deploy = flatpak_dir_load_deployed (dir, ref, NULL, NULL, NULL);
+      if (deploy == NULL)
+        continue;
+
+      origin = flatpak_dir_get_origin (dir, ref, NULL, NULL);
+      if (origin == NULL)
+        continue;
```

Comment:
I think the above is a bit inefficient. We first do flatpak_dir_load_deployed() which finds the deploy dir, loads and parses the metadata file, and then loads any overrides. Then flatpak_dir_get_origin() find the deploy dir again, loads the deploy data file and returns the origin from that.

Furthermore, the parsing of the metadata is only needed to the the runtime and sdk of the app, and that is already available in the deploy data (actually, not sdk but we can easily add it) so we should be able to completely skip the metadata file.

Reply:
I guess this is later deleted, but the load_deployed below still exists in the final code.